protocol/messenger_profile_showcase.go

```diff
@@ -1,88 +1,342 @@
+func toProfileShowcaseAssetProto(preferences []*ProfileShowcaseAssetPreference, visibility ProfileShowcaseVisibility) []*protobuf.ProfileShowcaseAsset {
+	assets := []*protobuf.ProfileShowcaseAsset{}
+	for _, preference := range preferences {
+		if preference.ShowcaseVisibility != visibility {
+			continue
 		}
+
+		assets = append(assets, &protobuf.ProfileShowcaseAsset{
+			Symbol: preference.Symbol,
+			Order:  uint32(preference.Order),
+		})
 	}
-	return nil
+	return assets
 }
 
-func (m *Messenger) SetProfileShowcasePreference(entry *ProfileShowcaseEntry) error {
-	return m.persistence.InsertOrUpdateProfileShowcasePreference(entry)
+func fromProfileShowcaseCommunityProto(messages []*protobuf.ProfileShowcaseCommunity) []*identity.ProfileShowcaseCommunity {
+	communities := []*identity.ProfileShowcaseCommunity{}
+	for _, entry := range messages {
+		communities = append(communities, &identity.ProfileShowcaseCommunity{
+			CommunityID: entry.CommunityId,
+			Order:       int(entry.Order),
+		})
+	}
+	return communities
 }
 
-func (m *Messenger) SetProfileShowcasePreferences(preferences ProfileShowcasePreferences) error {
-	err := preferences.Validate()
+func fromProfileShowcaseAccountProto(messages []*protobuf.ProfileShowcaseAccount) []*identity.ProfileShowcaseAccount {
+	accounts := []*identity.ProfileShowcaseAccount{}
+	for _, entry := range messages {
+		accounts = append(accounts, &identity.ProfileShowcaseAccount{
+			Address: entry.Address,
+			Name:    entry.Name,
+			ColorID: entry.ColorId,
+			Emoji:   entry.Emoji,
+			Order:   int(entry.Order),
+		})
+	}
+	return accounts
+}
+
+func fromProfileShowcaseCollectibleProto(messages []*protobuf.ProfileShowcaseCollectible) []*identity.ProfileShowcaseCollectible {
+	collectibles := []*identity.ProfileShowcaseCollectible{}
+	for _, entry := range messages {
+		collectibles = append(collectibles, &identity.ProfileShowcaseCollectible{
+			UID:   entry.Uid,
+			Order: int(entry.Order),
+		})
+	}
+	return collectibles
+}
+
+func fromProfileShowcaseAssetProto(messages []*protobuf.ProfileShowcaseAsset) []*identity.ProfileShowcaseAsset {
+	assets := []*identity.ProfileShowcaseAsset{}
+	for _, entry := range messages {
+		assets = append(assets, &identity.ProfileShowcaseAsset{
+			Symbol: entry.Symbol,
+			Order:  int(entry.Order),
+		})
+	}
+	return assets
+}
+
+func (m *Messenger) SetProfileShowcasePreferences(preferences *ProfileShowcasePreferences) error {
+	err := m.persistence.SaveProfileShowcasePreferences(preferences)
 	if err != nil {
 		return err
 	}
 
-	allPreferences := []*ProfileShowcaseEntry{}
+	return m.publishContactCode()
+}
+
+func (m *Messenger) GetProfileShowcasePreferences() (*ProfileShowcasePreferences, error) {
+	return m.persistence.GetProfileShowcasePreferences()
+}
+
+func (m *Messenger) EncryptProfileShowcaseEntriesWithContactPubKeys(entries *protobuf.ProfileShowcaseEntries, contacts []*Contact) (*protobuf.ProfileShowcaseEntriesEncrypted, error) {
```

Comment:
The list of contacts might be quite big actually, maybe pass it by pointer?

Reply:
it's a slice, so it's already a pointer